wasm_node_api/wasm_node_api.cc

```diff
@@ -0,0 +1,886 @@
+struct is_native<Native<T>> : public std::true_type {};
+
+template <typename T>
+constexpr auto is_native_v = is_native<T>::value;
+
+template<typename>
+struct is_wasm : public std::false_type {};
+
+template<typename T>
+struct is_wasm<Wasm<T>> : public std::true_type {};
+
+template <typename T>
+constexpr auto is_wasm_v = is_wasm<T>::value;
+
+template<class K>
+using kind = typename K::type;
+
+template<typename T>
+constexpr w::ValKind wasmType() {
+  if constexpr (std::is_same<kind<T>, float>::value) {
+    return w::ValKind::F32;
+  } else if constexpr (std::is_same<kind<T>, double>::value) {
+    return w::ValKind::F64;
+  #if __WORDSIZE == 64
+  } else if constexpr (is_native_v<T> && std::is_pointer<kind<T>>::value) {
+    return w::ValKind::I64;
+  #endif
+  } else {
+    return w::ValKind::I32;
+  }
+}
+
+template<w::ValKind... Args>
+constexpr auto exportVoid(callback cb) -> w::Func {
+  return w::Func(
+    w::FuncType({Args...}, {}),
+    cb
+  );
+};
+
+template<w::ValKind Return, w::ValKind... Args>
+constexpr auto exportFn(callback cb) -> w::Func {
+  return w::Func(
+    w::FuncType({Args...}, {Return}),
+    cb
+  );
+};
+
+template<typename Return, typename... Args>
+constexpr auto exportNative(callback cb) -> w::Func {
+  if constexpr(std::is_void<kind<Return>>::value) {
+    return exportVoid<wasmType<Args>()...>(cb);
+  } else {
+    return exportFn<wasmType<Return>(), wasmType<Args>()...>(cb);
+  }
+}
+
+template<typename Arg>
+constexpr Arg napiCast(const w::Val& v) {
+  if constexpr(std::is_same<Arg, float>::value) {
+    return (Arg) v.f32();
+  } else if constexpr(std::is_same<Arg, double>::value) {
+    return (Arg) v.f64();
+  } else {
+    return (Arg) v.i32();
+  }
+}
+
+template<typename Result>
+constexpr kind<Result> napiUnwrap(const w::Context* ctx, const w::Val& v) {
+  if constexpr(is_wasm_v<Result>) {
+    if constexpr(/*std::is_same<kind<Result>,
+                              v8::MaybeLocal<v8::Function>>::value*/
+                 is_function_pointer_v<kind<Result>>) {
+      return (kind<Result>) v.i32();/*(ctx->table->get(v.i32()));*/
+    } else {
+      return (kind<Result>) (&ctx->memory->data()[(uint32_t) v.i32()]);
+    }
+  #if __WORDSIZE == 64
+  } else if constexpr(std::is_pointer<kind<Result>>::value) { // Native pointer
+    return (kind<Result>) v.i64();
+  #endif
+  } else {
+    return napiCast<kind<Result>>(v);
+  }
+};
+
+template<typename Result>
+constexpr w::Val napiWrap(kind<Result> v) {
+  if constexpr(std::is_same<kind<Result>, float>::value) {
+    return w::Val((float) v);
+  } else if constexpr (std::is_same<kind<Result>, double>::value) {
+    return w::Val((double) v);
+  #if __WORDSIZE == 64
+  } else if constexpr(std::is_pointer<kind<Result>>::value) { // Native pointer
+      return w::Val(static_cast<int64_t>((size_t) v));
+  #endif
+  } else {
+    return w::Val(static_cast<int32_t>((size_t) v));
+  }
+}
+
+template<typename Result, typename... Args>
+struct napiApply {
```

Comment:
This function is analogous to `std::apply` in that it allows us to call a generic function using an array of arguments but is specifically designed to unwrap arguments using `napiUnwrap`. This allows us to statically create WASM call wrappers for N-API functions (i.e. converting every parameter into a N-API value before making the call).

Reply:
Part of this should be converted to a code comment.